prusti-specs/src/extern_spec_rewriter.rs

```diff
@@ -201,76 +136,470 @@ fn rewrite_method_inputs(item_ty: &syn::Type, method: &mut ImplItemMethod) ->
+                    return Err(syn::Error::new(
+                        item_method.default.as_ref().unwrap().span(),
+                        "Default methods in external trait specs are disallowed \
+                        because external specs are always trusted",
+                    ));
+                }
+            }
+        }
+
+        // Generics not supported
+        if item_trait.generics.params.len() > 0 {
+            return Err(syn::Error::new(
+                item_trait.span(),
+                "Generics in external trait specs are not supported",
+            ));
+        }
+
+        return Ok(());
+    }
+
+    /// Given a trait declaration, returns the bounds to `Self` in the where clause
+    /// # Example
+    /// ```
+    /// trait Foo where Self: Bar<Baz> + Qux, Self: Quux
+    /// ```
+    /// Returns `Bar<Baz> + Qux + Quux`
+    fn get_bounds_of_self(
+        item_trait: &syn::ItemTrait,
+    ) -> syn::punctuated::Punctuated<syn::TypeParamBound, syn::Token![+]> {
+        let mut bounds: syn::punctuated::Punctuated<syn::TypeParamBound, syn::Token![+]> =
+            syn::punctuated::Punctuated::new();
+
+        if let Some(ref where_clause) = item_trait.generics.where_clause {
+            for where_predicate in where_clause.predicates.iter() {
+                if let syn::WherePredicate::Type(where_predicate_type) = where_predicate {
+                    if let syn::Type::Path(bounded_type_path) = &where_predicate_type.bounded_ty {
+                        let path_ident = bounded_type_path.path.get_ident();
+                        if path_ident.is_some() && path_ident.unwrap() == "Self" {
+                            bounds.extend(where_predicate_type.bounds.clone());
+                        }
+                    }
+                }
+            }
+        }
+        bounds
+    }
+
+    /// Responsible for generating a struct
+    fn generate_new_struct(item_trait: &syn::ItemTrait) -> syn::Result<GeneratedStruct> {
+        let trait_ident = &item_trait.ident;
+
+        let name_generator = NameGenerator::new();
+        let struct_name = name_generator.generate_struct_name_for_trait(item_trait);
+        let struct_ident = syn::Ident::new(&struct_name, item_trait.span());
+
+        let mut new_struct: syn::ItemStruct = parse_quote_spanned! {item_trait.span()=>
+            #[allow(non_camel_case_types)] struct #struct_ident {}
         };
-        syn::parse2::<syn::GenericArgument>(ts).unwrap()
-    }).collect();
-    syn::parse_quote!{ < #(#args),* > }
+        // Find associated types in trait
+        let mut assoc_types_to_generics_map = HashMap::new();
+        let associated_type_decls = get_associated_types(item_trait);
+        for associated_type_decl in associated_type_decls {
+            let associated_type_ident = &associated_type_decl.ident;
+            let generic_ident = syn::Ident::new(
+                format!("T_AssocType_{}", associated_type_ident).as_str(),
```

Comment:
Unlikely, but this (and `T_Self` more so) might still have a name clash with user-declared types. `PrustiT_...`?